chain/near/src/chain.rs

```diff
@@ -0,0 +1,383 @@
+        _adapter: &TriggersAdapter,
+        filter: &TriggerFilter,
+    ) -> Result<BlockStreamEvent<Chain>, FirehoseError> {
+        let step = bstream::ForkStep::from_i32(response.step).unwrap_or_else(|| {
+            panic!(
+                "unknown step i32 value {}, maybe you forgot update & re-regenerate the protobuf definitions?",
+                response.step
+            )
+        });
+        let any_block = response
+            .block
+            .as_ref()
+            .expect("block payload information should always be present");
+
+        // Right now, this is done in all cases but in reality, with how the BlockStreamEvent::Revert
+        // is defined right now, only block hash and block number is necessary. However, this information
+        // is not part of the actual bstream::BlockResponseV2 payload. As such, we need to decode the full
+        // block which is useless.
+        //
+        // Check about adding basic information about the block in the bstream::BlockResponseV2 or maybe
+        // define a slimmed down stuct that would decode only a few fields and ignore all the rest.
+        let block = codec::BlockWrapper::decode(any_block.value.as_ref())?;
+
+        match step {
+            bstream::ForkStep::StepNew => Ok(BlockStreamEvent::ProcessBlock(
+                self.firehose_triggers_in_block(&block, filter)?,
+                Some(response.cursor.clone()),
+            )),
+
+            bstream::ForkStep::StepUndo => {
+                let block = block.block.as_ref().unwrap();
+                let header = block.header.as_ref().unwrap();
+
+                Ok(BlockStreamEvent::Revert(
+                    BlockPtr {
+                        hash: BlockHash::from(
+                            // FIXME (NEAR): Are we able to avoid the clone? I kind of doubt but worth checking deeper
+                            header.hash.as_ref().unwrap().bytes.clone(),
+                        ),
+                        number: header.height as i32,
+                    },
+                    Some(response.cursor.clone()),
+                ))
+            }
+
+            bstream::ForkStep::StepIrreversible => {
+                panic!("irreversible step is not handled and should not be requested in the Firehose request")
+            }
+
+            bstream::ForkStep::StepUnknown => {
+                panic!("unknown step should not happen in the Firehose response")
+            }
+        }
+    }
+}
+
+impl FirehoseMapper {
+    // FIXME: This should be replaced by using the `TriggersAdapter` struct directly. However, the TriggersAdapter trait
+    //        is async. It's actual async usage is done inside a manual `poll` implementation in `firehose_block_stream#poll_next`
+    //        value. An upcoming improvement will be to remove this `poll_next`. Once the refactor occurs, this should be
+    //        removed and TriggersAdapter::triggers_in_block should be use straight.
```

Comment:
Looking forward to this refactor.

Reply:
So do I. The thing that is a premise of that would be to remove the manual `Poll` implementation in `firehose_block_stream`. I checked the `reduce` something you mentioned, it has problems as far as I remember, will need to give it a second shot.

I refrain of refactoring it because the need to call yet another `Future` and manually resolve adding a bunch of state branches to the `Poll` state.